### ISSUE TITLE
Feature proposal: descending sequences

### ISSUE THREAD

**Issue by sferik** [crystal-lang/crystal#1391]
Quoting from the Ruby documentation:

> Ranges can be constructed using any objects that can be compared using the `<=>` operator. Methods that treat the range as a sequence expect the begin object to implement a `succ` method to return the next object in sequence.

In Ruby, the interface to implement a sequence is to define spaceship (`<=>`) and `succ` instance methods. I’d propose expanding that interface in Crystal to include `pred` to allow for descending sequences. Currently, in both Ruby and Crystal, if the start of the range is greater than the end of the range, the range is always empty.

``` crystal
(1..5).to_a #= [1, 2, 3, 4, 5]
(5..1).to_a #= []
```

Other languages (e.g. Haskell) allow ranges of descending elements.

``` haskell
[5, 4..1] -- [5,4,3,2,1]
```

Ruby and Crystal both already support `downto` enumerations, which use `pred`.

``` crystal
5.downto(1).to_a #= [5, 4, 3, 2, 1]
```

And it’s, of course, possible to reverse a range after it’s been converted to an array, but this is inefficient.

``` crystal
(1..5).to_a.reverse #= [5, 4, 3, 2, 1]
```

This feature is one of the things I miss in Ruby. Unfortunately, Ruby is too popular to make this change. Luckily, Crystal is not so popular yet, so I hope you will consider implementing this feature. :wink: :four_leaf_clover: 

Although this is a deviation from Ruby’s semantics, I believe it makes the language more beautiful by making it more symmetric—like a good :gem: should be.

**Comment by vyp** [comment]
+1 I agree with the last line.

**Comment by waterlink** [comment]
+1 for this feature

--
Best Regards,
Alexey Fedorov,
Sr Ruby, Clojure, Crystal, Golang Developer,
Microservices Backend Engineer,
+49 15757 486 476

2015-09-06 11:47 GMT+02:00 vyp <notifications@github.com>:

**Comment by asterite** [comment]
What about this:

```crystal
ary = [1, 2, 3, 4, 5]
ary[0 .. 2]   #=> [1, 2, 3]
ary[3 .. 2]   #=> Maybe [4, 3] ?
ary[0 .. -1] #=> [1, 2, 3, 4, 5]
ary[3 .. -4] #=> [] or [4, 3, 2] ?
```

I also think ranges are strange in that they only go upwards, but this change would be a huge change. Sometimes you do `1 .. n` to get values from 1 to n, but if `n` will happen to be zero you won't get what you'd expect... at least from Ruby. You'd have to do `1.upto(n)`, which is OK, but then you can't use ranges to iterate them safely. Or yes, we could have `each_upwards` and `each_downwards`, but then you won't have them like Enumerable (unless you use those without a block).

This is an interesting feature, we need to discuss it more :-)

**Comment by asterite** [comment]
I don't think we are going to do this, mostly because of my last comment, where current ranges are now descending but don't mean that. It will bring a lot of confusion. A descending sequence can be added as an external library if needed.

**Comment by straight-shoota** [comment]
I think this discussion needs a revisit.

The formal definition of `Range` as streching from `begin` to `end` does not imply `begin < end`. Descending ranges are valid have always been supported: `5..1` is completely valid. How such a range is interpreted, in say `Indexable#[]` is a different story.

The semantics of a descending range as an index accessor argument as demonstrated in https://github.com/crystal-lang/crystal/issues/1391#issuecomment-138080781 is defined by `Array` (or rather `Indexable`), not `Range`. Whether it is treated as empty in that context is orthogonal to the discussion about basic features on `Range` itself.

The change proposed here is about `Range` methods such as `(5..1).to_a`. IMO that should very obviously return `[5, 4, 3, 2, 1, 0]` instead of an empty array.
With #10209 `Range#step` already works like that because it delegates to `Number#step`.

I think other `Range` methods should do the same. Most notably basic `#each` and `#reverse_each`.
Potentially, they could also be refactored to use `#step` and migrate `Range` implementation from `#succ`/`#pred` interface to `#step`.

**Comment by oprypin** [comment]
If someone wrote this code:

```c
for (int i = 0; i <= n; ++i) {
    print(i);
}
```

do you think they'd be happy if the code had the ability to somehow print
```
0
-1
-2
```
?

Well, neither would the people who wrote
```crystal
(0..n).each do |i|
  puts i
end
```

This is a big breaking change introducing subtle bugs into all code, current and future.

Big "no" from me.

**Comment by asterite** [comment]
Same "no" from me to.

As I said before, if we make `(3..1).to_a` be `[3, 2, 1]` then people will expect `array[3..1]` to be the same as `[array[3], array[2], array[1]]` but that's not how it works.

**Comment by WhereIsX** [comment]
I think regardless of whether we want to support this behavior/feature, we should explicitly document that Range `begin < end`

**Comment by straight-shoota** [comment]
@oprypin The `for` example has a explicit `++` operator, so it would clearly be impossible to decrease. In Crystal that loop would be comparable to explicitly specifying the direction for example like `0.upto(n)`, `n.times` or `(0..n).step(1)`.

I would generally expect `(0..n).each` to iterate from the beginning (`0`) to end (`n`) of the range. 

If `n > 0` that means it iterates increasingly. If `n < 0` it means iterating decreasingly.

I'm don't have a reasonable explanation why `n < 0` would be interpreted as "do not iterate at all". As if no one in their right mind would possibly want to do a crazy thing like counting backwards.

The disparaty between `(a..b).step` and `a.step(to: b)` is really annoying as those methods are intended for the same purpose.

```cr
(5..1).step.to_a   # => []
5.step(to: 1).to_a # => [5, 4, 3, 2, 1]
```

Using `Range#step` vs. `Number#step` is mainly a question of whether the boundaries are passed as a variable, for example as method arguments and thus `Range` serves as a container for them.

Now, if I want to step a range generically, I can't use `Range#step` directly because it automatically defaults to increasing steps und thus only supports half the use cases.

Yes, this is a breaking change. But impact is probably very low.

Stdlib specs only break for the ones which test this behaviour explicitly in `range_spec.cr`. And after looking at a few shards, I found a single breaking spec in `pg` shard with `PG::Numeric#to_s`.

If you want to test it yourself, a patch is available at https://github.com/straight-shoota/crystal/tree/feature/range-descending-0.35.1 and a self-contained example for copy & paste at https://carc.in/#/r/a8o1
The patch is based on 0.35.1 instead of master in order to work with shards that don't compile because of other changes in master.
